tests/blockchain/blockchain_test_utils.py

```diff
@@ -0,0 +1,142 @@
+            # at this height that's also in chain. That would be an invariant
+            # violation
+            if row[1]:
+                # make sure we don't have any duplicate heights. Each block
+                # height can only have a single block with in_main_chain set
+                assert height not in in_chain
+                in_chain.add(height)
+                if height > max_height:
+                    max_height = height
+
+        # make sure every height is represented in the set
+        assert len(in_chain) == max_height + 1
+
+
+async def _validate_and_add_block(
+    blockchain: Blockchain,
+    block: FullBlock,
+    expected_result: Optional[ReceiveBlockResult] = None,
+    expected_error: Optional[Err] = None,
+    skip_prevalidation: bool = False,
+) -> None:
+    # Tries to validate and add the block, and checks that there are no errors in the process and that the
+    # block is added to the peak.
+    # If expected_result is not None, that result will be enforced.
+    # If expected_error is not None, that error will be enforced (unless prevalidation fails, in which
+    # case we return with no errors). If expected_error is not None, receive_block must return Err.INVALID_BLOCK.
+    # If expected_result == INVALID_BLOCK but expected_error is None, we will allow for errors to happen
+
+    await check_block_store_invariant(blockchain)
+    if skip_prevalidation:
+        results = PreValidationResult(None, uint64(1), None, False)
+    else:
+        pre_validation_results: Optional[
+            List[PreValidationResult]
+        ] = await blockchain.pre_validate_blocks_multiprocessing([block], {}, validate_signatures=False)
+
+        if pre_validation_results is None:
+            # Returning None from pre validation means an error occurred
+            if expected_error is not None or expected_result == ReceiveBlockResult.INVALID_BLOCK:
+                return None
+            else:
+                raise ValueError("Prevalidation returned None")
+        results = pre_validation_results[0]
+    if results.error is not None:
+        if expected_result == ReceiveBlockResult.INVALID_BLOCK:
+            await check_block_store_invariant(blockchain)
+            return None
+        if expected_error is None:
+            raise ValueError(Err(results.error))
+        elif Err(results.error) != expected_error:
+            raise ValueError(f"Expected {expected_error} but got {Err(results.error)}")
+        await check_block_store_invariant(blockchain)
+        return None
+
+    result, err, _, _ = await blockchain.receive_block(block, results)
+    await check_block_store_invariant(blockchain)
+    if expected_error is None and expected_result != ReceiveBlockResult.INVALID_BLOCK:
+        # Not expecting any errors
+        if err is not None:
+            # Got an error
+            raise ValueError(err)
+    else:
+        # Expecting an error
+        if err != expected_error:
```

Comment:
Not necessarily. We might expect error A but we got error B. So that is not what we expected so the test should fail. 
The if statement handles the normal case where no errors are expected.

Reply:
right. if I'm reading the code correctly, that's exactly what this condition checks. Ensuring the expected error is exactly equal to the error we got.

What I'm saying is that I would expect this check to always be performed, even when
```
expected_error is None and expected_result != ReceiveBlockResult.INVALID_BLOCK
```

Right now, you don't check this in that case, which looks suspicious. Maybe it's correct, but it's not as obvious as if you would just always ensure we got the error (or lack thereof) as we expected.

Reply:
Oh yeah that is intentional, to simplify test code. I can say for example:

`_validate_and_add_block(blockchain, b, expected_result=ReceiveBlockResult.INVALID_BLOCK)`

The comment at the top of the function explains this. I will also add a comment at the corresponding place.